File: solutions/855/main.py
from sortedcontainers import SortedList

class ExamRoom:
    # (-1, n) means seats between -1 and n can be used
    # it's open range 
    def __init__(self, n: int): 
        def dist(l, r): 
            if l == -1 or r == n:
                return r - l - 1  
            else:
                return (r - l) // 2 
        self.n = n 
        self.sl = SortedList(key=lambda x: (-dist(*x), x[0]))
        # record occupied seats to the left and right for removal 
        self.left = {}
        self.right = {}
        
        self.add((-1, n))

    def add(self, x) -> None: 
        self.sl.add(x)
        self.left[x[1]] = x[0]
        self.right[x[0]] = x[1]

    def delete(self, x) -> None: 
        self.sl.remove(x)
        self.left.pop(x[1])
        self.right.pop(x[0])

    def seat(self) -> int:
        x = self.sl[0]
        pick = (x[0]+x[1]) // 2 
        if x[0] == -1: 
            pick = 0 
        elif x[1] == self.n: 
            pick = self.n - 1 
        self.delete(x)
        self.add((x[0], pick))
        self.add((pick, x[1]))
        return pick

    def leave(self, p: int) -> None:
        l, r = self.left[p], self.right[p]
        self.delete((l, p))
        self.delete((p, r))
        self.add((l, r))

File: solutions/855/main.rs
use std::collections::BTreeSet;

struct ExamRoom {
    seats: BTreeSet<i32>, 
    capacity: i32, 
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl ExamRoom {

    fn new(n: i32) -> Self {
        ExamRoom {
            seats : BTreeSet::new(),
            capacity: n,
        }
    }
    
    fn seat(&mut self) -> i32 {
        // check if len is 0 
        if self.seats.len() == 0 {
            self.seats.insert(0); 
            return 0; 
        }

        let (mut left, mut right) = (0, 0); 
        let mut max_dist = i32::MIN; 
        let mut pre = 0; 
        let mut pick = 0; 
        for (i, item) in self.seats.iter().enumerate() {
            // handle first seat 
            if i == 0 && *item != 0 {
                if *item - pre > max_dist {
                    left = pre; 
                    right = *item; 
                    max_dist = right - left; 
                }
            } else {
                if (*item - pre) / 2 > max_dist {
                    left = pre; 
                    right = *item; 
                    max_dist = (right - left) / 2; 
                    pick = left + max_dist; 
                }
            }
            pre = *item; 
        }
        // handle last seat 
        if pre != self.capacity - 1 {
            if self.capacity - 1 - pre > max_dist {
                left = pre; 
                right = self.capacity - 1; 
                max_dist = right - left; 
                pick = right; 
            }
        }
        self.seats.insert(pick); 
        return pick;  
    }
    
    fn leave(&mut self, p: i32) {
        self.seats.remove(&p); 
    }
}

/**
 * Your ExamRoom object will be instantiated and called as such:
 * let obj = ExamRoom::new(n);
 * let ret_1: i32 = obj.seat();
 * obj.leave(p);
 */